model/src/lib.rs

```diff
@@ -52,26 +58,97 @@ pub struct OrderCreation {
+        hash_data[44..64].copy_from_slice(self.sell_token.as_fixed_bytes());
+        hash_data[76..96].copy_from_slice(self.buy_token.as_fixed_bytes());
+        self.sell_amount.to_big_endian(&mut hash_data[96..128]);
+        self.buy_amount.to_big_endian(&mut hash_data[128..160]);
+        hash_data[188..192].copy_from_slice(&self.valid_to.to_be_bytes());
+        hash_data[220..224].copy_from_slice(&self.app_data.to_be_bytes());
+        self.fee_amount.to_big_endian(&mut hash_data[224..256]);
+        hash_data[287] = match self.order_kind {
+            OrderKind::Sell => 0,
+            OrderKind::Buy => 1,
+        };
+        hash_data[319] = self.partially_fillable as u8;
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest_typed_data(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        let mut hash_data = [0u8; 66];
+        hash_data[0..2].copy_from_slice(&[0x19, 0x01]);
+        hash_data[2..34].copy_from_slice(domain_separator);
+        hash_data[34..66].copy_from_slice(&self.order_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest_message(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        let mut hash_data = [0u8; 92];
+        hash_data[0..28].copy_from_slice(b"\x19Ethereum Signed Message:\n64");
+        hash_data[28..60].copy_from_slice(domain_separator);
+        hash_data[60..92].copy_from_slice(&self.order_digest());
+        signing::keccak256(&hash_data)
+    }
+
+    pub fn signing_digest(&self, domain_separator: &[u8; 32]) -> [u8; 32] {
+        if self.signature.v & 0x80 == 0 {
+            self.signing_digest_typed_data(domain_separator)
+        } else {
+            self.signing_digest_message(domain_separator)
+        }
     }
-    pub fn order_digest(&self) -> H256 {
-        // dummy implementation, waiting for Valentins PR
-        H256::zero()
+
+    // If signature is valid returns the owner.
+    pub fn validate_signature(&self, domain_separator: &[u8; 32]) -> Option<H160> {
+        let v = self.signature.v & 0x1f;
+        let message = self.signing_digest(domain_separator);
+        let recovery = Recovery::new(message, v as u64, self.signature.r, self.signature.s);
+        let (signature, recovery_id) = recovery.as_signature()?;
+        signing::recover(&message, &signature, recovery_id).ok()
     }
-    pub fn order_uid(&self) -> OrderUid {
+
+    pub fn uid(&self, owner: &H160) -> OrderUid {
         let mut uid = OrderUid([0u8; 56]);
-        uid.0[0..32].copy_from_slice(self.order_digest().as_fixed_bytes());
-        uid.0[32..52].copy_from_slice(self.order_owner().as_fixed_bytes());
+        uid.0[0..32].copy_from_slice(&self.order_digest());
+        uid.0[32..52].copy_from_slice(owner.as_fixed_bytes());
         uid.0[52..56].copy_from_slice(&self.valid_to.to_be_bytes());
         uid
     }
 }
 
+// Intended to be used by tests that need signed orders.
+impl OrderCreation {
```

Comment:
nit: mark it to be only compiled under under test?

Reply:
No because it wouldn't exist for the tests of dependent crates.